homeassistant/components/tplink/entity.py

```diff
@@ -57,24 +95,213 @@ async def _async_wrap(self: _T, *args: _P.args, **kwargs: _P.kwargs) -> None:
-        self._attr_unique_id = device.device_id
+        self._device: Device = device
+        self._feature = feature
+
+        registry_device = device
+        name = device.alias
+        if parent and parent.device_type != Device.Type.Hub:
+            if not feature or feature.category == Feature.Category.Primary:
+                # Entity will be added to parent if not a hub and no feature parameter
+                # (i.e. core platform like Light, Fan) or feature is primary like state
+                registry_device = parent
+                name = registry_device.alias
+                self._attr_name = device.alias
+            else:
+                # Prefix the device name with the parent name unless it is a hub attached device.
+                # Sensible default for child devices like strip plugs or the ks240 where the child
+                # alias makes more sense in the context of the parent.
+                # i.e. Hall Ceiling Fan & Bedroom Ceiling Fan; Child device aliases will be Ceiling Fan
+                # and Dimmer Switch for both so should be distinguished by the parent name.
+                name = f"{parent.alias} {device.alias}"
+
         self._attr_device_info = DeviceInfo(
-            connections={(dr.CONNECTION_NETWORK_MAC, device.mac)},
-            identifiers={(DOMAIN, str(device.device_id))},
+            identifiers={(DOMAIN, str(registry_device.device_id))},
             manufacturer="TP-Link",
-            model=device.model,
-            name=device.alias,
-            sw_version=device.hw_info["sw_ver"],
-            hw_version=device.hw_info["hw_ver"],
+            model=registry_device.model,
+            name=name,
+            sw_version=registry_device.hw_info["sw_ver"],
+            hw_version=registry_device.hw_info["hw_ver"],
         )
+
+        if parent is not None and parent != registry_device:
+            self._attr_device_info["via_device"] = (DOMAIN, parent.device_id)
+        else:
+            self._attr_device_info["connections"] = {
+                (dr.CONNECTION_NETWORK_MAC, device.mac)
+            }
+
+        self._attr_unique_id = self._get_unique_id()
+        self._attr_entity_category = self._category_for_feature(feature)
+
+    def _get_unique_id(self) -> str:
+        """Return unique ID for the entity."""
+        device = self._device
+        if self._feature is not None:
+            feature_id = self._feature.id
+            # Special handling for primary state attribute (backwards compat for the main switch).
+            if feature_id == PRIMARY_STATE_ID:
+                return legacy_device_id(device)
+
+            return f"{legacy_device_id(device)}_{feature_id}"
+
+        # Light entities were handled historically differently
+        if isinstance(self, LightEntity):
+            unique_id = device.mac.replace(":", "").upper()
+            # For backwards compat with pyHS100
+            if device.device_type == DeviceType.Dimmer:
```

Comment:
If these are always Enums and not StrEnum we can use `is` instead of `==`

Reply:
Fixed